### PR TITLE
Feat/vanilla ae refactor

### PR DESCRIPTION
**Pull request by ab93**
- Refactor Vanilla AE
- Support static filters in Druid
- Add Percentile Scaler

### CODE REVIEW

File: numalogic/transforms/_movavg.py [hunk 1]
```diff
@@ -86,24 +87,20 @@ class ExpMovingAverage(StatelessTransformer):
     Args:
```
Comment (s0nicboOm) [inline numalogic/transforms/_movavg.py:87]: I think the class doc has to be updated for this function.

File: numalogic/transforms/_movavg.py [hunk 2]
```diff
@@ -114,32 +111,6 @@ def transform(self, input_: npt.NDArray[float], **__):
...
-
-        # Elemental multiply with beta powers
-        exp_avg = alpha * np.multiply(beta_powers, theta_beta_ratio)
-        if not self.bias_correction:
-            return exp_avg
-
-        # Calculate array of 1 / (1 - beta**i) values
-        return np.divide(exp_avg, 1.0 - beta_powers)
+        x_df = pd.DataFrame(input_, dtype=np.float32, copy=True)
+        x_smoothed = x_df.ewm(alpha=self.alpha).mean().to_numpy(dtype=np.float32)
```
Comment (s0nicboOm) [inline numalogic/transforms/_movavg.py:115]: I think we should return without a mean.

File: numalogic/backtest/_prom.py [hunk 3]
```diff
@@ -253,12 +265,17 @@ def generate_scores(
 
             winscores = postproc_udf.compute_feature_scores(
                 raw_scores[idx], self.nlconf.score.window_agg
+            )  # (nfeat,)
+
+            unified_raw_scores[idx] = postproc_udf.compute_unified_score(
+                winscores,
+                AggregatorConf(method=AggMethod.MEAN),
```
Comment (s0nicboOm) [inline numalogic/backtest/_prom.py:272]: Don't we want this config driven?

**Review by s0nicboOm — COMMENTED** [review]
minor comments

File: numalogic/udfs/postprocess.py [hunk 4]
```diff
@@ -122,7 +122,11 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
             load_latest=LOAD_LATEST,
             vertex=self._vtx,
         )
-        postproc_tx = self.postproc_factory.get_instance(postprocess_cfg)
+        postproc_tx = (
+            self.postproc_factory.get_instance(postprocess_cfg) if postprocess_cfg else None
+        )
+        if not postproc_tx:
```
Comment (gulshan02) [inline numalogic/udfs/postprocess.py:128]: should we raise error/exception here if postproc model is not present?

File: numalogic/transforms/_scaler.py [hunk 5]
```diff
@@ -69,3 +72,70 @@ def fit_transform(self, x: npt.NDArray[float], y=None, **_) -> npt.NDArray[float
     def _check_if_constant(self, x: npt.NDArray[float]) -> None:
         delta = np.max(x, axis=0) - np.min(x, axis=0)
         self._std[delta < self._eps] = 1.0
+
+
+class PercentileScaler(BaseTransformer):
```
Comment (gulshan02) [inline numalogic/transforms/_scaler.py:77]: can we add unit tests for the fit function in test_transforms.py

File: numalogic/transforms/_postprocess.py [hunk 6]
```diff
@@ -48,3 +48,13 @@ def __init__(self, scale_factor=10, smooth_factor=10):
...
         return tanh_norm(input_, scale_factor=self.scale_factor, smooth_factor=self.smooth_factor)
+
+
+class SigmoidNorm(StatelessTransformer):
+    def __init__(self, scale_factor: float = 10.0, smooth_factor: float = 0.5):
+        super().__init__()
+        self.scale_factor = scale_factor
+        self.smooth_factor = smooth_factor
+
+    def transform(self, x: npt.NDArray[float], **__) -> npt.NDArray[float]:
```
Comment (gulshan02) [inline numalogic/transforms/_postprocess.py:59]: can we add unit test for this one too

File: numalogic/udfs/postprocess.py (hunk 4 above)
Comment (s0nicboOm) [reply numalogic/udfs/postprocess.py:128]: I added that debug log. So, for some cases, we do not want to have postprocess model. I don't think that is an error. It is the expected behavior that we want :)

File: numalogic/transforms/_movavg.py (hunk 1 above)
Comment (gulshan02) [reply numalogic/transforms/_movavg.py:87]: +1 on updating class doc

File: numalogic/models/threshold/_static.py [hunk 7]
```diff
@@ -121,4 +121,6 @@ def score_samples(self, x: npt.NDArray[float]) -> npt.NDArray[float]:
         with values being anomaly scores.
         """
         self._validate_input(x)
-        return self.score_limit / (1.0 + np.exp(-self.coeff * (x.copy() - self.upper_limits)))
+        exp_arg = -self.coeff * (x - self.upper_limits)
+        exp_arg = np.clip(exp_arg, -88.72, 88.72)
```
[the remark this answers is not in the document]
Comment (ab93) [reply numalogic/models/threshold/_static.py:125]: to prevent floating overflow warning

File: numalogic/transforms/_movavg.py (hunk 2 above)
Comment (ab93) [reply numalogic/transforms/_movavg.py:115]: df.ewm does not return a real valued number. It just returns an object. We need to aggregate in the end.

File: numalogic/models/threshold/_median.py [hunk 8]
```diff
@@ -45,6 +59,19 @@ def _validate_input(x: npt.NDArray[float]) -> None:
     def fit(self, x: npt.NDArray[float]) -> Self:
         self._validate_input(x)
         self._thresh = np.percentile(x, self._max_percentile, axis=0)
+
+        if self._adjust_threshold:
+            for idx, _ in enumerate(self._thresh):
+                ratio = self._thresh[idx] / self._min_thresh
+                if ratio < 1e-2:
```
[the remark this answers is not in the document]
Comment (ab93) [reply numalogic/models/threshold/_median.py:66]: We can actually remove this, and revert back to what it was before.

File: numalogic/backtest/_prom.py (hunk 3 above)
Comment (ab93) [reply numalogic/backtest/_prom.py]: Yes it should be. Good catch.

File: numalogic/transforms/_movavg.py (hunk 1 above)
Comment (ab93) [reply numalogic/transforms/_movavg.py:84]: Done

File: numalogic/transforms/_movavg.py (hunk 2 above)
Comment (s0nicboOm) [reply numalogic/transforms/_movavg.py:112]: gotcha!

**Review by s0nicboOm — APPROVED** [review]
lgtm!

**Review by gulshan02 — APPROVED** [review]
lgtm

**Pull request merged**